proposals/texture-component-swizzle.md

```diff
@@ -0,0 +1,93 @@
+
+```webidl
+partial dictionary GPUTextureViewDescriptor {
+    GPUTextureComponentSwizzle swizzle;
+};
+
+// Structure specifying a custom color component mapping for a texture view.
+dictionary GPUTextureComponentSwizzle {
+    GPUComponentSwizzle r = "identity";
+    GPUComponentSwizzle g = "identity";
+    GPUComponentSwizzle b = "identity";
+    GPUComponentSwizzle a = "identity";
+};
+
+// A set of options to choose from when specifying a component swizzle.
+enum GPUComponentSwizzle {
+    "identity", // Set to the identity swizzle.
+    "zero",     // Force its value to 0.
+    "one",      // Force its value to 1.
+    "r",        // Take its value from the red channel of the texture.
+    "g",        // Take its value from the green channel of the texture.
+    "b",        // Take its value from the blue channel of the texture.
+    "a",        // Take its value from the alpha channel of the texture.
+};
+```
+
+The `GPUTexture.createView(descriptor)` algorithm is extended with the following validation rules changes:
+
+- If `descriptor.usage` includes the `RENDER_ATTACHMENT` or `STORAGE_BINDING` bit:
+  - `descriptor.swizzle.r` must be `"identity"`.
+  - `descriptor.swizzle.g` must be `"identity"`.
+  - `descriptor.swizzle.b` must be `"identity"`.
+  - `descriptor.swizzle.a` must be `"identity"`.
+
+## Javascript example
+
+Imagine you have a texture with a single red channel (`r8unorm` format). In your shader, you want to use this as a grayscale image where the red, green, and blue components all have the same value from the texture's red channel, and the alpha is fully opaque:
+
+```js
+const adapter = await navigator.gpu.requestAdapter();
+if (!adapter.features.has("texture-component-swizzle")) {
+  throw new Error("Texture component swizzle is not available");
+}
+// Explicitly request texture component mapping support.
+const device = await adapter.requestDevice({
+  requiredFeatures: ["texture-component-swizzle"],
+});
+
+// ... Assuming myTexture is a GPUTexture with a single red channel.
+
+const textureView = myTexture.createView({
+  swizzle: {
+    r: 'r',  // Map the view's red component to the texture's red channel
+    g: 'r',  // Map the view's green component to the texture's red channel
+    b: 'r',  // Map the view's blue component to the texture's red channel
+    a: 'one' // Force the view's alpha component to 1.0
+  }
+});
+```
+
+## Open Questions
+
+- Are there new validation rules needed if the view is depth-stencil or multisampled?
+
+## Resources
+
+- [Vkswizzle(3) Vulkan Manual Page](https://registry.khronos.org/vulkan/specs/latest/man/html/VkComponentMapping.html)
+- [D3D12_SHADER_RESOURCE_VIEW_DESC structure (d3d12.h)](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_shader_resource_view_desc)
+- [MTLTextureSwizzleChannels | Apple Developer Documentation](https://developer.apple.com/documentation/metal/mtltextureswizzlechannels)
```

Comment:
+ - [OpenGL ES 3.1 section 14.2.1](https://registry.khronos.org/OpenGL/specs/es/3.1/es_spec_3.1.pdf)

?

Reply:
Thanks @greggman! I've opened https://github.com/gpuweb/gpuweb/pull/5222

Reply:
Can we also add a compat rule that swizzle is taken into account when checking that all views for a draw / dispatch are compatible?